interface.go

```diff
@@ -0,0 +1,92 @@
+//
+// Copyright (c) 2019 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package registry
+
+import (
+	"github.com/pelletier/go-toml"
+)
+
+type Client interface {
+	// Registers the current service with Registry for discover and health check
+	Register() error
+
+	// Checks to see if the Registry contains the service's configuration.
+	CheckConfigurationExistence() (bool, error)
+
+	// Puts a full toml configuration into the Registry
+	PutConfiguration(configuration *toml.Tree, overwrite bool) error
+
+	// Gets the full configuration from Consul into the target configuration struct.
+	// Passed in struct is only a reference for Registry. Empty struct is fine
+	// Returns the configuration in the target struct as interface{}, which caller must cast
+	GetConfiguration(configStruct interface{}) (interface{}, error)
+
+	// Sets up a Consul watch for the target key and send back updates on the update channel.
+	// Passed in struct is only a reference for Registry, empty struct is ok
+	// Sends the configuration in the target struct as interface{} on updateChannel, which caller must cast
+	WatchForChanges(updateChannel chan<- interface{}, errorChannel chan<- error, configuration interface{}, waitKey string)
+
+	// Simply checks if Registry is up and running at the configured URL
+	IsRegistryRunning() bool
+
+	// Checks if a configuration value exists in the Registry
+	ConfigurationValueExists(name string) (bool, error)
+
+	// Gets a specific configuration value from the Registry
+	GetConfigurationValue(name string) ([]byte, error)
+
+	// Puts a specific configuration value into the Registry
+	PutConfigurationValue(name string, value []byte) error
+
+	// Gets the service endpoint information for the target ID from the Registry
+	GetServiceEndpoint(serviceId string) (ServiceEndpoint, error)
+
+	// Checks with the Registry if the target service is available, i.e. registered and healthy
+	IsServiceAvailable(serviceID string) bool
+}
```

Comment:
This would require all the services to treat them separately, i.e. ConfigRegistry and ServiceRegistry. Correct? 

An implementation with two different providers could implement this single interface using the two different providers in the appropriate functions and then the services still treat as one Registry.

Reply:
Yeah, this is like the discussion we've been having the Security WG about whether we should have a facade that wraps both configuration and secrets. I think doing so is an SRP violation. Not going to die on this hill, but calling it out that we may need to refactor at a later date.